pkg/oauthserver/authenticator/password/keystonepassword/keystonepassword.go

```diff
@@ -16,24 +18,116 @@ import (
+}
+
+type keystoneToken struct {
+	Methods    []string      `json:methods`
+	ExpiresAt  string        `json:expires_at`
+	User       keystoneUser  `json:user`
+	AuditIDS   []string      `json:audit_ids`
+	IssuedAt   string        `json:issued_at`
+}
+
+type keystoneResponseBody struct {
+	Token  keystoneToken  `json:token`
+}
+
 // keystonePasswordAuthenticator uses OpenStack keystone to authenticate a user by password
 type keystonePasswordAuthenticator struct {
-	providerName   string
-	url            string
-	client         *http.Client
-	domainName     string
-	identityMapper authapi.UserIdentityMapper
+	providerName           string
+	url                    string
+	client                 *http.Client
+	domainName             string
+	identityMapper         authapi.UserIdentityMapper
+	useKeystoneIdentity    bool
 }
 
 // New creates a new password authenticator that uses OpenStack keystone to authenticate a user by password
 // A custom transport can be provided (typically to customize TLS options like trusted roots or present a client certificate).
 // If no transport is provided, http.DefaultTransport is used
-func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper) authenticator.Password {
+func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper, useKeystoneIdentity bool) authenticator.Password {
 	if transport == nil {
 		transport = http.DefaultTransport
 	}
 	client := &http.Client{Transport: transport}
-	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper}
+	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper, useKeystoneIdentity}
+}
+
+// Authenticate user and return his ID from Keystone
+func getUserIDv3(client *gophercloud.ProviderClient, options gophercloud.AuthOptions) (string, error) {
+	// Override the generated service endpoint with the one returned by the version endpoint.
+	v3Client := openstack.NewIdentityV3(client)
+
+	// copy the auth options to a local variable that we can change. `options`
+	// needs to stay as-is for reauth purposes
+	v3Options := options
+
+	var scope *tokens3.Scope
+	if options.TenantID != "" {
+		scope = &tokens3.Scope{
+			ProjectID: options.TenantID,
+		}
+		v3Options.TenantID = ""
+		v3Options.TenantName = ""
+	} else {
+		if options.TenantName != "" {
+			scope = &tokens3.Scope{
+				ProjectName: options.TenantName,
+				DomainID:    options.DomainID,
+				DomainName:  options.DomainName,
+			}
+			v3Options.TenantName = ""
+		}
+	}
+
+	result := tokens3.Create(v3Client, v3Options, scope)
+
+	token, err := result.ExtractToken()
+	if err != nil {
+		return "", err
+	}
+
+	catalog, err := result.ExtractServiceCatalog()
+	if err != nil {
+		return "", err
+	}
+
+	// Parse the body to obtain userID from the response
+	out, err := json.Marshal(result.Body)
```

Comment:
`err := json.NewDecoder(result.Body).Unmarshal(&bd)` can be done directly

Reply:
I started using mapstructure lib: https://github.com/mitchellh/mapstructure
They recommend it as a more advanced way to decode json. I hope it is not a big issue.

result.Body is an interface, so we also have to convert a type, which makes code not so beautiful as with mapstructure: "err = mapstructure.Decode(result.Body, &body)"